File: testdata/fixtures/python/safe/log_safe.py
# Safe: Structured logging with sanitized input fields
# Expected: No findings for BATOU-LOG-001

import logging
import re
from flask import Flask, request

app = Flask(__name__)
logger = logging.getLogger(__name__)


def sanitize_log(value):
    """Remove newlines and control characters from log input."""
    return re.sub(r"[\r\n\t]", "_", str(value))[:200]


@app.route("/login", methods=["POST"])
def login():
    username = sanitize_log(request.form.get("username", ""))
    action = sanitize_log(request.form.get("action", ""))

    # Safe: parameterized logging with sanitized values
    logger.info("Login attempt: username=%s action=%s ip=%s",
                username, action, request.remote_addr)

    logger.warning("Failed login: username=%s ip=%s",
                   username, request.remote_addr)
    return "OK"


@app.route("/search")
def search():
    query = sanitize_log(request.args.get("q", ""))
    # Safe: structured log format, sanitized input
    logger.info("Search query: q=%s", query)
    return "OK"

File: testdata/fixtures/go/safe/log_safe.go
package safe

import (
	"log/slog"
	"net/http"
	"strings"
)

// SAFE: Structured logging with sanitized input fields.
// Should NOT trigger BATOU-LOG-001 or any log injection rules.

func HandleLogin(w http.ResponseWriter, r *http.Request) {
	username := sanitizeLogField(r.FormValue("username"))
	action := sanitizeLogField(r.FormValue("action"))

	// Safe: structured logging with sanitized values
	slog.Info("login attempt",
		"username", username,
		"action", action,
		"ip", r.RemoteAddr,
	)

	slog.Warn("failed login",
		"username", username,
		"ip", r.RemoteAddr,
	)
}

func sanitizeLogField(input string) string {
	// Remove newlines and control characters to prevent log injection
	s := strings.ReplaceAll(input, "\n", "_")
	s = strings.ReplaceAll(s, "\r", "_")
	s = strings.ReplaceAll(s, "\t", "_")
	if len(s) > 200 {
		s = s[:200]
	}
	return s
}
